tck/src/main/java/org/eclipse/microprofile/config/tck/PropertyExpressionsTest.java

```diff
@@ -0,0 +1,190 @@
+
+        assertThrows(NoSuchElementException.class, () -> config.getValue("expression", String.class));
+    }
+
+    @Test
+    public void noExpressionComposed() {
+        Config config = buildConfig("expression", "${my.prop${compose}}");
+
+        assertThrows(NoSuchElementException.class, () -> config.getValue("expression", String.class));
+    }
+
+    @Test
+    public void multipleExpansions() {
+        Config config = buildConfig("my.prop", "1234", "my.prop.two", "${my.prop}", "my.prop.three",
+                                            "${my.prop.two}", "my.prop.four", "${my.prop.three}");
+
+        assertEquals("1234", config.getValue("my.prop", String.class));
+        assertEquals("1234", config.getValue("my.prop.two", String.class));
+        assertEquals("1234", config.getValue("my.prop.three", String.class));
+        assertEquals("1234", config.getValue("my.prop.four", String.class));
+    }
+
+    @Test
+    public void infiniteExpansion() {
+        Config config = buildConfig("my.prop", "${my.prop}");
+
+        assertThrows(IllegalArgumentException.class, () -> config.getValue("my.prop", String.class));
+    }
+
+    @Test
+    public void withoutExpansion() {
+        Config config = buildConfig("my.prop", "1234", "expression", "${my.prop}", "mp.config.property.expressions", "false");
+
+        assertEquals("${my.prop}", config.getValue("expression", String.class));
+    }
+
+    @Test
+    public void escape() {
+        assertEquals("${my.prop}", buildConfig("expression", "\\${my.prop}").getValue("expression", String.class));
+    }
+
+    @Test
+    void arrayEscapes() {
+        Config config = buildConfig("list", "cat,dog,${mouse},sea\\,turtle", "mouse", "mouse");
+
+        final List<String> list = config.getValues("list", String.class);
+        assertEquals(4, list.size());
+        assertEquals(list, Stream.of("cat", "dog", "mouse", "sea,turtle").collect(toList()));
+    }
+
+    @Test
+    void escapeBraces() {
+        Config config = buildConfig("my.prop", "${value:111{111}");
+        assertEquals("111{111", config.getValue("my.prop", String.class));
+    }
+
+    @Test
+    void expressionMissing() {
+        Config config = buildConfig("my.prop", "${expression}", "my.prop.partial", "${expression}partial");
+
+        assertThrows(Exception.class, () -> config.getValue("my.prop", String.class));
+        assertThrows(Exception.class, () -> config.getValue("my.prop.partial", String.class));
+    }
+
+    private static Config buildConfig(String... keyValues) {
+        if (keyValues.length % 2 != 0) {
+            throw new IllegalArgumentException("keyValues array must be a multiple of 2");
+        }
+
+        Map<String, String> properties = new HashMap<>();
+        for (int i = 0; i < keyValues.length; i += 2) {
+            properties.put(keyValues[i], keyValues[i + 1]);
+        }
+
+        return ConfigProviderResolver.instance().getBuilder()
```

Comment:
I took a further look at the tests. You do need to test CDI, which is why I asked.

@Inject @ConfigProperty(name="url") SomeType t;
By the way, please also add another test to ensure the defaultValue can contain ${}, which will be expanded as well.

Reply:
As talked in the call with @jbee @radcortez, the spec will state default value will not perform expression substituation.